### DISCUSSION TITLE
Embedded attachments don't show

### DISCUSSION THREAD

**Discussion by ShadeBlack** [dompdf/dompdf#2962] [Q&A] [answered]
I can't tell if this is an issue or if I'm doing something wrong...

I have a pdf that includes images in the pdf html, but these show as thumbnails and obviously cannot include full resolution.
Embeds are the choice - I can do a zip if I can't get the embed to work, but embed would be preferred.

I couldn't find an up-to-date example of the embed on the wiki, so this is what I've got:
```
// tmp file array
$tfa = [
  ['file' => 'tmp\fa_91D9.tmp', 'name' => 'fa_91D9.tmp', 'desc' => 'added: 23-Jul-2022 0503'],
  ['file' => 'tmp\fa_91D9.tmp', 'name' => 'fa_91D9.tmp', 'desc' => 'added: 23-Jul-2022 0503'],
  ['file' => 'tmp\fa_91D9.tmp', 'name' => 'fa_91D9.tmp', 'desc' => 'added: 23-Jul-2022 0503'],
];

// dompdf
$dompdf = new Dompdf();
$cpdf = new Cpdf();
$dompdf->loadHtml($html);
$dompdf->setPaper('A4', 'portrait');
$dompdf->render();

// attachments
foreach ($tfa as $k => $v) {
	$vf = $v["file"] ?? null;
	$vn = $v["name"] ?? null;
	$vd = $v["desc"] ?? null;
	debug([$vf, $vn, $vd]);
	$cpdf->addEmbeddedFile(
		$vf,
		$vn,
		$vd
	);
}

$dompdf->stream("$filename.pdf");

```

Note about my attachments:
I grab them from an external storage server, 
create tmpnam, 
put contents,
add to $tfa

After I did this, I have no php errors, and the file size of the pdf generated increased to the expected size given the files included.
However, when I load the pdf into Adobe Reader / Firefox, the attachments are nowhere to be found.

Does anyone have any insight why the attachments might not show?

**Comment by bsweeney** [comment, marked as answer]
Instead of creating a new Cpdf instance you should use the one created by Dompdf:

```
$cpdf = $dompdf->getCanvas()->get_cpdf();
```

**Reply by ShadeBlack** [reply]
Hi bsweeney
Thanks for helping!

I removed the `new cpdf();`
and added: `$cpdf = $dompdf->getCanvas()->get_cpdf();` in its place

So this is the example I saw in one of the issues after I searched.
my ide says get_cpdf is undefined method,
and get_canvas is deprecated.

dompdf.php has getCanvas so I switched to that.
CPDF.php actually does have get_cpdf, and running it produces no issues.

So I dumped $cpdf after the addEmbeddedFile to see:
$cpdf => objects => [my attachments]
![image](https://user-images.githubusercontent.com/10591504/180625257-3a93f0ff-16a8-40da-92ab-84e8788d49f5.png)

so $cpdf has the information about the attachments in it's instance, 
but still my output pdf still does not seem to include them.

For reference, I am running dompdf 2.0.0
installed via composer

**Reply by bsweeney** [reply]
Sorry for the deprecated method use. That's what I get for straight up using some copy-pasta.

`get_cpdf` likely shows up as undefined in your IDE because it's not part of the Canvas adapter interface. You can give it a type hint to let it know you're using the CPDF adapter if you want to clear that notice.

I did a quick test and it seems fine. Do you have a sample of your output?

**Reply by ShadeBlack** [reply]
no problem: just had to cut out the storage url, but nothing else changed in this file:
[VPI - 24-Jul-2022 2007.pdf](https://github.com/dompdf/dompdf/files/9176505/VPI.-.24-Jul-2022.2007.pdf)
I used Firefox and Adobe reader to look for the attachments.  As far as I'm aware, Chrome / Edge doesn't have an attachment viewer.
_There shouldn't be any personal/sensitive info in this pdf._

If you've got a working test file, Could I possibly look at that?  It might show what I'm doing incorrectly.

**Reply by bsweeney** [reply]
For my working test I just mirrored the sample you provided (array of file info processed after calling render) so I'm not doing anything extra. The only difference being that I am calling the `output` method instead of the `stream` method. That shouldn't matter, though, since `stream` is just a wrapper around `output`.

The PDF itself probably wouldn't help since your sample does appear to legitimately be missing the files.

The fact that Cpdf indicates the embedded files exist in the object collection does indicate that embedded files are working correctly. I believe you should be able to address the missing files in the output by retrieving the Canvas instance later in your script, after calling the `render` method. Dompdf will [replace the current Canvas instance](https://github.com/dompdf/dompdf/blob/v2.0.0/src/Dompdf.php#L720-L723) when the page size is different from the default (something I hope to address by implementing #613).

**Reply by ShadeBlack** [reply]
Incredible!  This appears to be exactly the issue for me!
Can now confirm that `render` kills the attachments.
I moved the `get_cpdf` to just before the attachment loop, and placed the `render` above it like so:

```
// instantiate
$dompdf = new Dompdf();
$dompdf->loadHtml($html);
$dompdf->setPaper('A4', 'portrait');
$dompdf->render();

// attachments
$cpdf = $dompdf->getCanvas()->get_cpdf();
foreach ($tfa as $k => $v) $cpdf->addEmbeddedFile($v["file"], $v["name"], $v["desc"]);

// output
$dompdf->stream("$filename.pdf");
```

File size increased, 
Attachments available in Firefox - can be saved / viewed.

I wish I had the foresight to play around a little more with the order of operations before bothering you with all this.  I will be sending a small donation your way as thanks! :)

**Reply by bsweeney** [reply]
Aw, thanks! Sorry for the trouble it's not an obvious issue. But you certainly went far in debugging (inspecting the Cpdf instance) which I appreciate.
